src/ws/cockpitwebservice.c

```diff
@@ -2126,148 +2133,138 @@ resource_response_new (CockpitWebService *self,
   return rr;
 }
 
-static gchar *
-pop_package_name (const gchar *path,
-                  const gchar **remaining_path)
-{
-  /*
-   * Parses packages in this form:
-   *
-   * /package/path/to/file.ext
-   *
-   * For the above will return 'package', and set remaining_path
-   * to point to /path/to/file.ext
-   */
-
-  const gchar *beg = NULL;
-
-  if (path && path[0] == '/')
-    {
-      beg = path + 1;
-      path = strchr (beg, '/');
-    }
-
-  if (!beg)
-    return NULL;
-
-  if (remaining_path)
-    *remaining_path = path;
-
-  if (path)
-    return g_strndup (beg, path - beg);
-  else
-    return g_strdup (beg);
-}
-
 static gboolean
 resource_respond (CockpitWebService *self,
                   GHashTable *headers,
                   CockpitWebResponse *response,
-                  const gchar *remaining_path)
+                  const gchar *where)
 {
   ResourceResponse *rr;
   CockpitSession *session = NULL;
   const gchar *host = NULL;
-  const gchar *name = NULL;
-  const gchar *path = NULL;
   gchar *package = NULL;
+  gchar *val = NULL;
+  gchar *language = NULL;
   gboolean ret = FALSE;
   GHashTableIter iter;
   GBytes *command;
   gchar **parts = NULL;
-  gchar **languages;
   JsonObject *object;
-  JsonArray *accept;
-  guint i;
+  JsonObject *heads;
+  gpointer key;
+  gpointer value;
 
-  package = pop_package_name (remaining_path, &path);
-  if (!package || !path)
+  if (where[0] == '@')
     {
-      g_debug ("invalid path: %s", remaining_path);
-      goto out;
-    }
-
-  /* Split a package@host name */
-  parts = g_strsplit (package, "@", 2);
-  name = parts[0];
-  host = parts[1];
-
-  /* No host specified? Always ask local first, faster */
-  if (host == NULL)
-    {
-      session = g_hash_table_lookup (self->sessions.by_host, "localhost");
-      if (session && session->packages)
-        {
-          if (g_hash_table_lookup (session->packages, name))
-            host = session->host;
-        }
+      host = where + 1;
     }
-
-  /* Now look through all the other hosts */
-  if (host == NULL)
+  else if (where[0] == '$')
     {
       g_hash_table_iter_init (&iter, self->sessions.by_transport);
       while (g_hash_table_iter_next (&iter, NULL, (gpointer *)&session))
         {
-          if (session->packages)
+          if (session->checksum && g_str_equal (session->checksum, where + 1))
             {
-              if (g_hash_table_lookup (session->packages, name))
-                {
-                  host = session->host;
-                  break;
-                }
+              host = session->host;
+              break;
             }
         }
     }
-
-  /* Default to local if we can't find it */
-  if (host == NULL)
+  else
     {
-      host = "localhost";
-      session = NULL;
+      goto out;
     }
 
-  if (!session)
-    session = lookup_or_open_session_for_host (self, host, NULL, self->creds, FALSE);
+  session = lookup_or_open_session_for_host (self, host, NULL, self->creds, FALSE);
```

Comment:
This might be reached with host == NULL.  That's what you meant with "$unknown automatically loads the resource from localhost", right?